mm2src/derives/skip_serializing_none/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+    where
+        F: Fn(&mut Field) -> Result<(), String>,
+    {
+        match fields {
+            Fields::Unit => Ok(()),
+            Fields::Named(ref mut fields) => fields
+                .named
+                .iter_mut()
+                .map(|field| function(field).map_err(|err| Error::new(field.span(), err)))
+                .collect_error(),
+            Fields::Unnamed(ref mut fields) => fields
+                .unnamed
+                .iter_mut()
+                .map(|field| function(field).map_err(|err| Error::new(field.span(), err)))
+                .collect_error(),
+        }
+    }
+    if let Ok(mut input) = syn::parse::<ItemStruct>(input.clone()) {
+        apply_on_fields(&mut input.fields, function)?;
+        Ok(quote!(#input))
+    } else if let Ok(mut input) = syn::parse::<ItemEnum>(input) {
+        input
+            .variants
+            .iter_mut()
+            .map(|variant| apply_on_fields(&mut variant.fields, function))
+            .collect_error()?;
+        Ok(quote!(#input))
+    } else {
+        Err(Error::new(
+            Span::call_site(),
+            "The attribute can only be applied to struct or enum definitions.",
+        ))
+    }
+}
+
+fn skip_serializing_none_add_attr_to_field(field: &mut Field) -> Result<(), String> {
+    if is_std_option(&field.ty) {
+        let has_skip_serializing_if = field_has_attribute(field, "serde", "skip_serializing_if");
+
+        let mut has_always_attr = false;
+        for attr in field.clone().attrs {
+            let has_attr = attr
+                .parse_meta()
+                .map_err(|e| e.to_string())?
+                .path()
+                .is_ident("serialize_always");
+
+            has_always_attr |= has_attr;
+
+            if !has_attr {
+                field.attrs.retain(|ele| *ele == attr);
+            }
+        }
+
+        if has_always_attr && has_skip_serializing_if {
+            let mut msg = r#"The attributes `serialize_always` and `serde(skip_serializing_if = "...")` cannot be used on the same field"#.to_string();
+            if let Some(ident) = &field.ident {
+                msg += ": `";
+                msg += &ident.to_string();
+                msg += "`";
+            }
+            msg += ".";
+            return Err(msg);
+        }
+
+        if has_skip_serializing_if || has_always_attr {
+            return Ok(());
+        }
+
+        let attr = parse_quote!(
+            #[serde(skip_serializing_if = "Option::is_none")]
+        );
+        field.attrs.push(attr);
+    } else {
+        for attr in field.attrs.iter() {
+            if attr
+                .parse_meta()
+                .map_err(|e| e.to_string())?
+                .path()
+                .is_ident("serialize_always")
+            {
+                return Err("`serialize_always` may only be used on fields of type `Option`.".into());
+            }
+        }
+    }
+    Ok(())
+}
```

Comment:
The below simplifies this function a bit and removes some of the nesting
```suggestion
fn is_serialize_always(attr: &syn::Attribute) -> Result<bool, String> {
    attr.parse_meta()
        .map_err(|e| e.to_string())
        .map(|meta| meta.path().is_ident("serialize_always"))
}

fn skip_serializing_none_add_attr_to_field(field: &mut Field) -> Result<(), String> {
    if !is_std_option(&field.ty) {
        for attr in field.attrs.iter() {
            if is_serialize_always(attr)? {
                return Err("`serialize_always` may only be used on fields of type `Option`.".into());
            }
        }

        return Ok(());
    }

    let has_skip_serializing_if = field_has_attribute(field, "serde", "skip_serializing_if");

    let mut has_always_attr = false;
    let mut attrs_to_retain = Vec::new();
    for attr in &field.attrs {
        let has_attr = is_serialize_always(attr)?;
        has_always_attr |= has_attr;
        if !has_attr {
            attrs_to_retain.push(attr.clone());
        }
    }
    field.attrs = attrs_to_retain;

    if has_always_attr && has_skip_serializing_if {
        let mut msg = r#"The attributes `serialize_always` and `serde(skip_serializing_if = "...")` cannot be used on the same field"#.to_string();
        if let Some(ident) = &field.ident {
            msg += ": `";
            msg += &ident.to_string();
            msg += "`";
        }
        msg += ".";
        return Err(msg);
    }

    if !has_skip_serializing_if && !has_always_attr {
        field.attrs.push(parse_quote!(
            #[serde(skip_serializing_if = "Option::is_none")]
        ));
    }

    Ok(())
}
```

Reply:
thank you!